File: pkg/proto/he.go
package proto

import (
	"context"
	"reflect"

	"github.com/complyue/anko/vm"
	"github.com/complyue/hbi/pkg/errors"
)

// HostingEnv is the container of HBI artifacts, including:
//   * functions
//   * object constructors (special functions taking n args, returning 1 object)
//   * reactor methods
//   * value objects
// These artifacts need to be explicitly exposed to a hosting environment,
// to accomodate landing of `peer-scripting-code` from the other end.
type HostingEnv struct {
	ve       *vm.Env
	po       *PostingEnd
	ho       *HostingEnd
	exposure []string
}

// NewHostingEnv creates a new, empty hosting environment.
func NewHostingEnv() *HostingEnv {
	he := &HostingEnv{
		ve:       vm.NewEnv(),
		exposure: make([]string, 0, 50),
	}
	return he
}

// AnkoEnv returns the underlying Anko (https://github.com/mattn/anko) env.
func (he *HostingEnv) AnkoEnv() *vm.Env {
	return he.ve
}

// Po reutrns the posting endpoint attached to this env.
func (he *HostingEnv) Po() *PostingEnd {
	return he.po
}

// Ho returns the hosting endpoint attached to this env.
func (he *HostingEnv) Ho() *HostingEnd {
	return he.ho
}

// ExposedNames returns a copy of the slice of names, ever exposed through the `Expose*()` calles,
// up to the point it is called.
func (he *HostingEnv) ExposedNames() []string {
	return append([]string(nil), he.exposure...)
}

// RunInEnv executes `code` within this environment.
func (he *HostingEnv) RunInEnv(code string, ctx context.Context) (result interface{}, err error) {
	// this needs not to be thread safe, should only be called from a single hosting goroutine
	result, err = he.ve.ExecuteContext(ctx, code)
	return
}

// NameExposed tells whether the specified name is exported from this env, up to the point it is called.
func (he *HostingEnv) NameExposed(name string) bool {
	// list of exposed names should not be quite long, linear search is okay
	for _, expName := range he.exposure {
		if expName == name {
			return true
		}
	}
	return false
}

func (he *HostingEnv) validateExposeName(name string) {
	if he.NameExposed(name) {
		panic(errors.Errorf("exposure name %s already used", name))
	}
}

// ExposeFunction exposes an arbitrary function to this env.
func (he *HostingEnv) ExposeFunction(name string, fun interface{}) {
	if reflect.TypeOf(fun).Kind() != reflect.Func {
		panic(errors.Errorf("not a function: %T = %#v", fun, fun))
	}
	he.validateExposeName(name)
	if err := he.ve.Define(name, fun); err != nil {
		panic(errors.RichError(err))
	}
	he.exposure = append(he.exposure, name)
}

// ExposeCtor exposes the specified constructor function,
// with expose name as the function's return type name,
// or `typeAlias` if it's not empty.
//
// note a constructor function should return one and only one value.
func (he *HostingEnv) ExposeCtor(ctorFunc interface{}, typeAlias string) {
	ft := reflect.TypeOf(ctorFunc)
	if ft.Kind() != reflect.Func {
		panic(errors.Errorf("not a function: %T = %#v", ctorFunc, ctorFunc))
	}
	if ft.NumOut() != 1 {
		panic(errors.Errorf("ctor function should return just 1 value, not %v", ft.NumOut()))
	}
	frt := ft.Out(0)
	if frt.Kind() == reflect.Ptr {
		frt = frt.Elem()
	}
	if typeAlias == "" {
		typeAlias = frt.Name()
	}
	he.validateExposeName(typeAlias)
	if err := he.ve.Define(typeAlias, ctorFunc); err != nil {
		panic(errors.RichError(err))
	}
	he.exposure = append(he.exposure, typeAlias)
}

// ExposeValue exposes the specified `val` with `name`.
func (he *HostingEnv) ExposeValue(name string, val interface{}) {
	he.validateExposeName(name)
	if err := he.ve.Define(name, val); err != nil {
		panic(errors.RichError(err))
	}
	he.exposure = append(he.exposure, name)
}

// Get returns what ever value associated with `name` in this env, which can be an:
//   * exposed artifact
//   * intrinsic object
//   * implanted artifact by peer script
// or nil if the name does not bind to anything in this env.
func (he *HostingEnv) Get(name string) interface{} {
	if val, err := he.ve.Get(name); err == nil {
		// undefined symbol, return nil instead of panic here
		return val
	}
	return nil
}

// ExposeReactor exposes all methods declared (by the list returned from `NamesToExpose`)
// from `reactor`. If `reactor` does not implements `Reactor`, all its exported methods are
// exposed.
func (he *HostingEnv) ExposeReactor(reactor interface{}) {
	var expNameWhiteList map[string]struct{}
	if reactorObj, ok := reactor.(Reactor); ok {
		wlNames := reactorObj.NamesToExpose()
		expNameWhiteList = make(map[string]struct{}, len(wlNames))
		for i := range wlNames {
			expNameWhiteList[wlNames[i]] = struct{}{}
		}
	}

	var expName string

	pv := reflect.ValueOf(reactor)
	pt := pv.Type()
	if pt.Kind() != reflect.Ptr {
		panic(errors.Errorf("pointer implementation expected for reactor, found type: %T", reactor))
	}
	cv := pv.Elem()
	ct := cv.Type()

	// expose exported fields of the context struct
	for fi, nf := 0, ct.NumField(); fi < nf; fi++ {
		sf := ct.Field(fi)
		fv := cv.Field(fi)
		if sf.PkgPath != "" {
			continue // ignore unexported field
		}
		if sf.Anonymous { // expose methods of embeded field struct
			he.ExposeReactor(fv.Addr().Interface())
			continue
		}
		if expNameWhiteList != nil {
			if _, ok := expNameWhiteList[sf.Name]; !ok {
				// not in declared exposure list
				continue
			}
		}
		// expose field getter/setter func
		expName = "Get" + sf.Name
		if err := he.ve.Define(expName, func() interface{} {
			return fv.Interface()
		}); err != nil {
			panic(errors.RichError(err))
		}
		if !he.NameExposed(expName) {
			he.exposure = append(he.exposure, expName)
		}
		expName = "Set" + sf.Name
		if err := he.ve.Define(expName, reflect.MakeFunc(
			reflect.FuncOf([]reflect.Type{sf.Type}, []reflect.Type{}, false),
			func(args []reflect.Value) (results []reflect.Value) {
				fv.Set(args[0])
				return
			},
		).Interface()); err != nil {
			panic(errors.RichError(err))
		}
		if !he.NameExposed(expName) {
			he.exposure = append(he.exposure, expName)
		}
	}

	// collect exported methods of the context struct
	for mi, nm := 0, pv.NumMethod(); mi < nm; mi++ {
		mt := pt.Method(mi)
		if mt.PkgPath != "" {
			continue // ignore unexported method
		}
		mv := pv.Method(mi)
		if expNameWhiteList != nil {
			if _, ok := expNameWhiteList[mt.Name]; !ok {
				// not in declared exposure list
				continue
			}
		}
		expName = mt.Name
		if err := he.ve.Define(expName, mv.Interface()); err != nil {
			panic(errors.RichError(err))
		}
		if !he.NameExposed(expName) {
			he.exposure = append(he.exposure, expName)
		}
	}

}

// Reactor is the interface optionally implemented by a type whose instances are
// to be exposed to a `HostingEnv` by calling `he.ExposeReactor()`.
//
// If a reactor object does not implement this interface, all its exported
// fields and methods will be exposed.
type Reactor interface {
	NamesToExpose() []string
}

File: pkg/proto/he.py
import ast
from typing import *
from ..log import *

from .po import PostingEnd
from .ho import HostingEnd

__all__ = ["HostingEnv", "run_py"]

logger = get_logger(__name__)


class HostingEnv:
    """
    HostingEnv is the container of hbi artifacts, including:
      * functions
      * object constructors (special functions taking n args, returning 1 object)
      * value objects
      * reactor methods
    These artifacts need to be explicitly exposed to a hosting environment,
    to accomodate landing of peer scripting code.

    """

    __slots__ = ("_globals", "_locals", "_exposed_names", "_po", "_ho")

    def __init__(self, open_: bool = True):
        self._globals = {}
        self._locals = self._globals if open_ else None
        self._exposed_names = []
        self._po = None
        self._ho = None

    @property
    def po(self) -> "PostingEnd":
        return self._po

    @property
    def ho(self) -> "HostingEnd":
        return self._ho

    @property
    def globals(self) -> dict:
        return self._globals

    @property
    def locals(self) -> Optional[dict]:
        return self._locals

    @property
    def exposed_names(self) -> Tuple[str]:
        return tuple(self._exposed_names)

    def run_in_env(self, code: str) -> object:
        globals_ = self._globals
        locals_ = self._locals
        if locals_ is None:
            # locals should not be None, create one to fire-and-forget
            locals_ = {}

        return run_py(code, globals_, locals_, "<hbi-code>")

    def name_exposed(self, name: str) -> bool:
        return name in self._exposed_names

    def _validate_expose_name(self, name: str):
        if name in self._exposed_names:
            raise AttributeError(f"name `{name!s}` already exposed")

    def expose_function(self, name: str, func: Callable):
        if not callable(func):
            raise TypeError(f"not a function: {func!r}")
        if name is None:
            name = func.__name__
        self._validate_expose_name(name)
        self._exposed_names.append(name)
        self._globals[name] = func

    def expose_ctor(self, ctor: type):
        if not callable(ctor):
            raise TypeError(f"constructor not callable: {ctor!r}")
        try:
            name = ctor.__name__
        except AttributeError:
            raise TypeError(f"constructor has no name: {ctor!r}")
        self._validate_expose_name(name)
        self._exposed_names.append(name)
        self._globals[name] = ctor

    def expose_value(self, name: str, value: object):
        self._validate_expose_name(name)
        self._exposed_names.append(name)
        self._globals[name] = value

    def get(self, name: str) -> object:
        return self._globals.get(name, None)

    def expose_reactor(self, reactor: object, cls: type = None):

        try:
            names_to_expose = (
                cls.names_to_expose if cls is not None else reactor.names_to_expose
            )
        except AttributeError:
            logger.warning(
                f"Reactor type {cls or type(reactor)!r} has no exposure declaration,"
                " all methods not starting with underscore will be exposed.",
                exc_info=True,
            )

            # expose all public (name not started with underscore) artifacts
            for name, _ in (cls or type(reactor)).__dict__.items():
                assert len(name) > 0, "?!"
                if "_" == name[0]:  # started with underscore, consider private/internal
                    continue
                if name in self._exposed_names:
                    # name already exposed, not to overwrite more specific attrs with a more general one
                    continue
                self._exposed_names.append(name)
                self._globals[name] = getattr(reactor, name)

            if cls is not None:
                # finished exposing a specific super class, done
                return

            # try expose inherited artifacts from super classes
            for cls in type(reactor).__mro__[1:]:
                if "builtins" == cls.__module__:
                    continue  # no do for builtin types
                self.expose_reactor(reactor, cls)

            return  # done without an exposure declarartion

        # with an exposure declaration, expose whatever-and-only names declared in the list
        for name in names_to_expose:  # AttributeError following will raise through
            expose_val = getattr(reactor, name)
            if name in self._exposed_names:
                # name already exposed, not to overwrite more specific attrs with a more general one
                continue
            self._exposed_names.append(name)
            self._globals[name] = expose_val


def run_py(
    code: str, globals_: dict, locals_: dict = None, src_name="<py-code>"
) -> object:
    """
    Run arbitrary Python code in supplied globals/locals, return evaluated value of last statement.

    """
    try:
        ast_ = ast.parse(code, src_name, "exec")
        last_expr = None
        last_def_name = None
        for field_ in ast.iter_fields(ast_):
            if "body" != field_[0]:
                continue
            if len(field_[1]) > 0:
                le = field_[1][-1]
                if isinstance(le, ast.Expr):
                    last_expr = ast.Expression()
                    last_expr.body = field_[1].pop().value
                elif isinstance(le, (ast.FunctionDef, ast.ClassDef)):
                    last_def_name = le.name
        exec(compile(ast_, src_name, "exec"), globals_, locals_)
        if last_expr is not None:
            return eval(compile(last_expr, src_name, "eval"), globals_, locals_)
        elif last_def_name is not None:
            return defs[last_def_name]
        return None
    except Exception:
        logger.error(
            f"""Error running HBI code:
-=-{src_name}-=-
{code!s}
-=*{src_name}*=-
""",
            exc_info=True,
        )
        raise
